exclude burnin in pps and edit on tests

curiosusJR opened the pull request:
Suggestion for #922

curiosusJR committed:
add test for dnPhyloBrownian

curiosusJR committed:
exclude burnin in pps

curiosusJR committed:
add burnin=0 in test_pps to fit the output

davidcerny reviewed (commented):
The fix itself looks okay, thanks. Could you briefly describe the purpose of the new test? It seems that `dnPhyloBrownian` is already covered by a bunch of other tests; presumably this one is trying to do something that the others don't?

curiosusJR commented:
Oh the new test is for #896 and I did not find `dnPhyloBrownian` in current 'tests/' directory:
```
╰─❯ rg  'dnPhyloBrownian' .
./test_BM/scripts/mcmc_BM.Rev
61:traits ~ dnPhyloBrownianREML(psi, branchRates=1.0, siteRates=sigma, nSites=1)

./test_BM_MVN/scripts/mcmc_BM_MVN.Rev
64:traits ~ dnPhyloBrownianMVN(psi, branchRates=sigma, siteRates=rate_scale, rootStates=root_state, nSites=1)

./test_pps/scripts/mcmc_BM_pps.Rev
52:X ~ dnPhyloBrownianREML(tree, branchRates=sqrt(sigma2) )
```
so I added these since  @davidcerny mentioned in Slack two months ago, sorry for the delay😅.

davidcerny commented:
@curiosusJR You're right – I only searched for `dnPhyloBrownian` rather than `dnPhyloBrownian(`, so I didn't catch the fact that all existing BM tests actually employ either `dnPhyloBrownianREML` or `dnPhyloBrownianMVN`.

bredelings commented on src/core/analysis/PosteriorPredictiveSimulation.cpp:
```diff
@@ -73,7 +73,8 @@ void RevBayesCore::PosteriorPredictiveSimulation::run( int thinning )
     
-    size_t n_samples = traces[0].size();
+    size_t burnin = traces[0].getBurnin();
+    size_t n_samples = traces[0].size() - burnin;
```
It seems like `sim_pid_start` specifies where the samples from this replicate begin.  So if we have burnin, we want to start looking at samples a bit after this.  So maybe we should change the line `size_t index_sample = sim_pid_start` to `size_t index_sample = sim_pid_start + burnin`

bredelings commented:
Hi @curiosusJR, can you please expand the cover letter a bit?  In order for me to follow the changes, it would be helpful if you could explain why the indexing is correct.

curiosusJR commented:
Sorry for any confusion caused.

The only modification was to reduce the value of `n_sample` by minus `burnin` and to skip the burnin portion of the trace when indexing into it. 

The original code in `src/core/analysis/PosteriorPredictiveSimulation.cpp:146`, index_sample begin from the first iteration to the end. After reduce the `n_sample`,  `index_sample` start from 0 and end to `traces[0].size() - burnin - 1`,  and what I want is from `burnin` to `traces[0].size() - 1` so I added `burnin` back to the `index_sample` to make it move. 

I also verified this behavior using a small toy dataset descibed in #922 but use a special `posterior.var`: 
```
Iteration	Replicate_ID	topology
0	0	(T_1[&index=1]:0.000000,(T_2[&index=2]:0.000000,T_4[&index=4]:0.000000)[&index=5]:0.000000,T_3[&index=3]:0.000000)[&index=6]:0.000000;
1	1	(T_1[&index=1]:0.000000,T_4[&index=4]:0.000000,(T_3[&index=3]:0.000000,T_2[&index=2]:0.000000)[&index=5]:0.000000)[&index=6]:0.000000;
2	0	(T_1[&index=1]:0.000000,(T_3[&index=3]:0.000000,T_2[&index=2]:0.000000)[&index=5]:0.000000,T_4[&index=4]:0.000000)[&index=6]:0.000000;
3	1	(T_4[&index=4]:0.000000,T_1[&index=1]:0.000000,(T_3[&index=3]:0.000000,T_2[&index=2]:0.000000)[&index=5]:0.000000)[&index=6]:0.000000;
4	0	(T_4[&index=4]:0.000000,T_1[&index=1]:0.000000,(T_2[&index=2]:0.000000,T_3[&index=3]:0.000000)[&index=5]:0.000000)[&index=6]:0.000000;
5	1	(T_1[&index=1]:5.900000,(T_3[&index=3]:0.000000,T_2[&index=2]:0.000000)[&index=5]:0.000000,T_4[&index=4]:0.000000)[&index=6]:0.000000;
6	0	(T_4[&index=4]:0.008145,T_1[&index=1]:0.117820,(T_3[&index=3]:0.160589,T_2[&index=2]:0.154287)[&index=5]:0.146046)[&index=6]:0.000000;
7	1	(T_1[&index=1]:0.340981,(T_3[&index=3]:0.087132,T_2[&index=2]:0.042477)[&index=5]:0.104404,T_4[&index=4]:0.000360)[&index=6]:0.000000;
8	0	((T_2[&index=2]:0.269447,T_3[&index=3]:0.135982)[&index=5]:0.071733,T_4[&index=4]:0.072248,T_1[&index=1]:0.614333)[&index=6]:0.000000;
9	1	(T_1[&index=1]:0.284499,(T_2[&index=2]:0.000000,T_3[&index=3]:0.000000)[&index=5]:0.048102,T_4[&index=4]:0.025435)[&index=6]:0.000000;
```
In this case, the tree length in the first 5 iterations, as the burnin, is zero so there should be no variation in simulated samples. I specify only T_1 is variable with branch length=5.9 in the sixth sample and T_2 should be the same with T_3 in the last sample for marking the order. 

I ran both current version of revbayes (output attached in `sims`) and modified version of revbayes (output attached in `sims_new`) on it.  [sims.tar.gz](https://github.com/user-attachments/files/24573266/sims.tar.gz

If the index is correct, the output from modified version of revbayes should be only 5 iterations and only T_1 is variable in the first iteration and T_2 should be the same with T_3 in the last iteration. In comparison, the currently version have 5 invariable iterations burnin in output. 

So that I confirmed that only post-burnin samples are accessed and used in the posterior predictive simulation. Based on this test, the indexing is correct and consistent with the intended semantics of the burnin parameter.

davidcerny commented:
I wonder if this solution (especially the change made to line 147) is general enough to correctly handle `thinning` values greater than 1. Unfortunately, I'm not able to run my own tests to find out, because the original issue doesn't provide a fully reproducible script (see #922), and neither does the zipped directory you shared here.

curiosusJR commented:
I considered this situation simply and I thought the `thinning` value would not change the burning generation numbers. I tested this with full script as shown: 
```
data <- readDiscreteCharacterData("input.nex")

n_taxa <- data.ntaxa()
taxa <- data.taxa()

moves    = VectorMoves()
Q := fnJC(4) 
topology ~ dnUniformTopologyBranchLength(taxa, branchLengthDistribution=dnExponential(10.0))
moves.append( mvNNI(topology, weight=n_taxa) )
moves.append( mvSPR(topology, weight=n_taxa) )
seq ~ dnPhyloCTMC(tree=topology,  Q=Q,  type="DNA")
seq.clamp(data)
mymodel = model(Q)
monitors = VectorMonitors()
monitors.append( mnScreen(printgen=1) )
monitors.append( mnStochasticVariable(filename="posterior.var",printgen=1) )
mymcmc = mcmc(mymodel, monitors, moves, nruns=1)
mymcmc.run(15)

trace = readStochasticVariableTrace("posterior.var", burnin=0.25, delimiter=TAB);
pps = posteriorPredictiveSimulation(mymodel, directory="sims", trace);
pps.run(2);

clear()
q()
```
In total 16 generations with `burnin=0.25` and `thinning=2` is set here. So it would be `16 x (1-0.25) / 2 = 6` simulated samples if I got it right. 

```
╰─❯ rm sims; rb run.Rb > /dev/null && ls sims/ | wc -l
8

╰─❯ rm sims; ../../projects/cmake/build/rb run.Rb > /dev/null && ls sims/ | wc -l
6
```

davidcerny reviewed (approved):
I can indeed confirm the interaction between `burnin` and `thinning` is as desired. Using your `posterior.var` file with 10 iterations indexed 0 through 9, and a modified version of your test script specifying a `burnin` of 0.2 and a `thinning` factor of 2, I get 10 * (1 − 0.2) / 2 = 4 simulated datasets based on iterations 2, 4, 6, and 8 – exactly the way it should be. Approving.
